instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/TracingRequestStreamHandler.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.awslambda.v1_0;
+
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.RequestStreamHandler;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.Tracer;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A base class similar to {@link RequestStreamHandler} but will automatically trace invocations of
+ * {@link #doHandleRequest(InputStream input, OutputStream output, Context)}.
+ */
+public abstract class TracingRequestStreamHandler implements RequestStreamHandler {
+
+  private final AwsLambdaTracer tracer;
+
+  /**
+   * Creates a new {@link TracingRequestStreamHandler} which traces using the default {@link
+   * Tracer}.
+   */
+  protected TracingRequestStreamHandler() {
+    this.tracer = new AwsLambdaTracer();
+  }
+
+  /**
+   * Creates a new {@link TracingRequestStreamHandler} which traces using the specified {@link
+   * Tracer}.
+   */
+  protected TracingRequestStreamHandler(Tracer tracer) {
+    this.tracer = new AwsLambdaTracer(tracer);
+  }
+
+  /**
+   * Creates a new {@link TracingRequestStreamHandler} which traces using the specified {@link
+   * AwsLambdaTracer}.
+   */
+  protected TracingRequestStreamHandler(AwsLambdaTracer tracer) {
+    this.tracer = tracer;
+  }
+
+  @Override
+  public final void handleRequest(InputStream input, OutputStream output, Context context)
+      throws IOException {
+    Span span = tracer.startSpan(context, Kind.SERVER);
+    Throwable error = null;
+    try (Scope ignored = tracer.startScope(span)) {
+      doHandleRequest(input, output, context);
```

Comment:
I think we need to wrap `OutputStream` so the span is closed on `OutputStream.close` which could be asynchronous?

Reply:
It's an option I considered. However now, with span issued on lambda exit, stream handler behaves exactly as "no stream" counterpart. In particular, the span gets send even if underlying output stream flush fails - I understand that in the case of RequestHandler implementations serialization / writing happens after the lambda exits. So with this implementation we have similar behaviour in both cases. What do you think?

Reply:
I don't know the details of lambda well but does it not allow something like

```
CompletableFuture<Foo> response = backend.getFoo();

response.thenAccept(writeToOutputStream);
```

If this is allowed than I think we need to make sure the span survives until the output is closed since that's the length of the request and how we generally model async requests.

Reply:
With this in mind, the idea totally makes sense. Will wrap OOS then :)